prereise/gather/demanddata/transportation_electrification/generate_scaling_factors.py

```diff
@@ -0,0 +1,206 @@
+        df_ua.squeeze().loc[lambda x: x != "[no data]"],
+        df_state.rename(index=state2abv).squeeze(),
+    )
+
+
+def calculate_vmt_for_ua(census_ua, tht_ua):
+    """Calculate the total annual Vehicle Miles Traveled (VMT) in urban areas
+
+    :param dict census_ua: dictionary as returned by :func:`load_census_ua`
+    :param pandas.Series tht_ua: vmt per capita in urban areas as returned by
+        :func:`load_dot_vmt_per_capita`
+    :return: (*dict*) -- keys are state abbreviation and values are data frames giving
+        annual vmt by urban areas.
+    """
+
+    tht_ua_format = tht_ua.copy()
+    vmt_for_ua = {}
+
+    pattern = [",", "-", ".", " "]
+    for p in pattern:
+        tht_ua_format.index = tht_ua_format.index.str.replace(p, "", regex=True)
+    format2original = dict(zip(tht_ua_format.index, tht_ua.index))
+
+    for s in census_ua:
+        census_ua_format = census_ua[s].copy()
+        for p in pattern:
+            census_ua_format.index = census_ua_format.index.str.replace(
+                p, "", regex=True
+            )
+        common = set(tht_ua_format.index).intersection(set(census_ua_format.index))
+        vmt_for_ua[s] = (
+            pd.DataFrame(
+                {
+                    "Annual VMT": [
+                        365 * tht_ua_format.loc[i] * census_ua_format.loc[i]
+                        for i in common
+                    ]
+                },
+                index=list(common),
+            )
+            .rename(index=format2original)
+            .squeeze()
+        )
+
+    return vmt_for_ua
+
+
+def calculate_vmt_for_state(census_state, tht_state):
+    """Calculate the total annual Vehicle Miles Traveled (VMT) in states
+
+    :param dict census_state: dictionary as returned by :func:`load_census_state`
+    :param pandas.Series tht_state: vmt per capita in states as returned by
+        :func:`load_dot_vmt_per_capita`
+    :return: (*pandas.Series*) -- Indices are state abbreviations and values are and
+        values are annual vmt in state.
+    """
+    common = set(tht_state.index).intersection(set(census_state.index))
+
+    vmt_for_state = pd.Series(
+        [tht_state.loc[i] * census_state.loc[i] for i in common], index=list(common)
+    )
+    return vmt_for_state
+
+
+def calculate_urban_rural_fraction(vmt_for_ua, vmt_for_state):
+    """Calculate the percentage of Vehicle Miles Traveled (VMT) in urban and rural areas
+
+    :param dict vmt_for_ua: dictionary as returned by :func:`calculate_vmt_for_ua`.
+    :param pandas.Series vmt_for_state: series as returned by
+        :func:`calculate_vmt_for_state`
+    """
+    vmt_for_ua_perc = vmt_for_ua.copy()
+    vmt_for_ra_perc = {}
+    for s in vmt_for_ua:
+        if s in vmt_for_state.index:
+            vmt_for_ua_perc[s] = vmt_for_ua[s] / vmt_for_state.loc[s]
+            vmt_for_ra_perc[s] = 1 - vmt_for_ua_perc[s].sum()
+        else:
+            vmt_for_ua_perc[s] = 1
+            vmt_for_ra_perc[s] = 0
```

Comment:
Maybe we could do `pd.Series({s:v.sum() for s, v in vmt_for_ua.items()})` then using series operations to get the results in one shot.

Reply:
Not sure I understand why we would want to sum over all urban areas in a state. At this point of the calculation we want to get, for each urban area in a state, the fraction of VMT with respect to the total in the state:
```
>>> vmt_ua_perc["AL"]
Birmingham, AL         0.144348
Pensacola, FL-AL       0.001210
Mobile, AL             0.060350
Florence, AL           0.011524
Tuscaloosa, AL         0.021803
Decatur, AL            0.006910
Anniston-Oxford, AL    0.015453
Dothan, AL             0.014312
Huntsville, AL         0.045015
Gadsden, AL            0.013453
Columbus, GA-AL        0.008350
Montgomery, AL         0.049128
Auburn, AL             0.013323
Name: Annual VMT, dtype: float64
```
and the percentage of VMT is simply the rmainder

Reply:
Either way, I was thinking `rural VMT = state total VMT - sum(each UA VMT within the state)`, so that we have the first term and second term aligned by state on the right hand side, which can be done without looping through the state. But yeah, it doesn't matter as long as it works;)